vowpalwabbit/nn.cc

```diff
@@ -169,7 +152,11 @@ void predict_or_learn_multi(nn& n, single_learner& base, example& ec)
   shared_data sd;
   memcpy(&sd, n.all->sd, sizeof(shared_data));
   {
-    sd_guard(n.all, &sd);
+    auto* saved_all = n.all;
+    auto* saved_sd = n.all->sd;
+    // guard for all.sd as it is modified - this will restore the state at the end of the scope.
+    auto sd_guard = VW::scope_guard([saved_all, saved_sd]() { saved_all->sd = saved_sd; });
```

Comment:
I'm not sure if I like this or if I hate it... On one hand its a generic guard, on the other it still requires a lot of manual setup that leaves a lot of room for mistakes

Reply:
I think this is because the old guard wasn't actually just a guard but it also handed setting up the state of the all->sd pointer to a new value. This separates the concerns to /just/ the cleanup action

Reply:
Guards typically handle the setup and teardown of the objects being passed in; think of a lock guard. The part I dislike is that you need to manually do the setup while the teardown is automated. It feels like you could get into a lot of trouble if they're not paired together

Reply:
Yeah I guess it is kind of a taste thing. If there is a pattern that enables reuse of a guard then a specific object may make sense - but for one offs it seems excessive to create lots of objects with custom behavior. Renaming it to scope_exit may be better?

This pattern is inspired by the WIL (Windows internal library) scope_exit functionality: https://github.com/microsoft/wil/blob/cd51fa9d3c24e9e9f5d4cdf1a9768d68f441fc48/include/wil/resource.h#L543